src/esp/gfx/TextRenderer.cpp

```diff
@@ -0,0 +1,141 @@
+  if (!Cr::Utility::Resource::hasGroup("fonts")) {
+    importFontResources();
+  }
+  CORRADE_ASSERT(
+      fontSize > 0,
+      "TextRenderer::TextRenderer: font size" << fontSize << "is illegal.", );
+  CORRADE_ASSERT(
+      cacheSize > 0,
+      "TextRenderer::TextRenderer: cache size" << cacheSize << "is illegal.", );
+
+  font_ = fontManager_.loadAndInstantiate("StbTrueTypeFont");
+  CORRADE_ASSERT(
+      font_,
+      "TextRenderer::TextRenderer: cannot load and instantiate the font", );
+
+  Cr::Utility::Resource rs("fonts");
+  std::string fontName = "SourceSansPro-SemiBold.ttf";
+  if (!font_->openData(rs.getRaw(fontName), fontSize)) {
+    Cr::Utility::Fatal{-1} << "TextRenderer::TextRenderer: The font" << fontName
+                           << "does not exist.";
+  }
+
+  font_->fillGlyphCache(cache_, characters);
+
+  updateAspectRatio(aspectRatio);
+}
+
+size_t TextRenderer::createRenderer(float onScreenCharacterSize,
+                                    size_t glyphCount,
+                                    TextLocation location) {
+  CORRADE_ASSERT(onScreenCharacterSize > 0,
+                 "TextRenderer::createRenderer: the character size"
+                     << onScreenCharacterSize << "is illegal.",
+                 0);
+
+  CORRADE_ASSERT(glyphCount > 0,
+                 "TextRenderer::createRenderer: the capacity" << glyphCount
+                                                              << "is illegal.",
+                 0);
+
+  Mn::Text::Alignment alignment = Mn::Text::Alignment::TopLeft;
+  ;
+  switch (location) {
+    case TOP_LEFT:
+      // has been initialized above
+      break;
+    case TOP_RIGHT:
+      alignment = Mn::Text::Alignment::TopRight;
+      break;
+    default:
+      Cr::Utility::Fatal{-1} << "TextRenderer::createRenderer: The location"
+                             << "is not supported.";
+      break;
+  }
+  text_.emplace_back(
+      Cr::Containers::Pointer<Mn::Text::Renderer2D>(new Mn::Text::Renderer2D(
+          *font_, cache_, onScreenCharacterSize, alignment)));
+  locations_.emplace_back(location);
+
+  size_t rendererId = text_.size() - 1;
+  text_[rendererId]->reserve(glyphCount, Mn::GL::BufferUsage::DynamicDraw,
+                             Mn::GL::BufferUsage::StaticDraw);
+
+  return rendererId;
+}
+
+void TextRenderer::updateText(const std::string& text, size_t rendererId) {
+  CORRADE_ASSERT(rendererId < text_.size(),
+                 "TextRenderer::updateText: the rendererId"
+                     << rendererId << "is out of range.", );
+  text_[rendererId]->render(text);
+}
+
+TextRenderer& TextRenderer::draw(const Mn::Color3& color, size_t rendererId) {
+  CORRADE_ASSERT(
+      rendererId < text_.size(),
+      "TextRenderer::draw: the rendererId" << rendererId << "is out of range.",
+      *this);
+  auto entry = locations_[rendererId];
+  textShader_
+      .setTransformationProjectionMatrix(
+          textProjection_ *
+          Mn::Matrix3::translation(
```

Comment:
> Why so complicated? Why not, for example, have the text projection equal to window size and the text size in pixels? That'll make sure the text size stays the same when you resize / maximize the window.

Sorry, I cannot understand the sentence: `have the text projection equal to window size and the text size in pixels?`. Based on what you said later, I assume you would like to display the text in fixed size (no matter what the window size is, the text size is fixed.) Can you give me the expression of the projection matrix? (this is the most straightforward way.)

If I understand you correctly, then what if the window size becomes tiny even smaller than the text size?

Reply:
I cleaned up and updated the text example to show this better -- please see https://github.com/mosra/magnum-examples/blob/91faa149da5def876558ea93d1ac47d3df0f37f3/src/text/TextExample.cpp#L105-L113

Reply:
I see. Thanks a lot.
The example code is much clearer than before.
When you say "mesh origin", I understand the "mesh" is for the 2D text, not the model in the scene. (I know it sounds silly, but for readers who have no ideas how the text are displayed using modern OpenGL, it can be confusing. They do not know there is a "mesh" for a character.)

Reply:
By *mesh* I mean the `GL::Mesh` that's populated by the `render()` function and later drawn, yes.